src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java

```diff
@@ -276,4 +285,41 @@ private <T> T runWithConnectionErrorCheck(CrashableSupplier<T, Exception> task)
             throw e;
         }
     }
+
+    /**
+     * Cleanup component, delete docker image when component being removed.
+     * @throws PackageLoadingException from getPackageRecipe
+     */
+    @Override
+    public void cleanup() throws PackageLoadingException, InvalidArtifactUriException, DockerImageDeleteException {
+        // this docker image not only used by itself
+        if (!ifImageUsedByOther(componentStore)) {
+            Image image = DockerImageArtifactParser
+                    .getImage(ComponentArtifact.builder().artifactUri(artifact.getArtifactUri()).build());
+            dockerClient.deleteImage(image);
+        }
+    }
+
+    /**
+     * @param componentStore componentStore
+     * @return true: this image used by other; false: not used.
+     * @throws PackageLoadingException from getPackageRecipe
+     */
+    public boolean ifImageUsedByOther(ComponentStore componentStore) throws PackageLoadingException {
+        Map<String, Set<String>> allVersions = componentStore.listAvailableComponentVersions();
+        for (Map.Entry<String, Set<String>> versions : allVersions.entrySet()) {
+            String compName = versions.getKey();
+            Set<String> localVersions = new HashSet<>(versions.getValue());
+            for (String compVersion : localVersions) {
+                ComponentIdentifier identifier = new ComponentIdentifier(compName, new Semver(compVersion));
```

Comment:
`new Semver` can throw an exception. Handle the exceptions that can happen here by logging, and continuing the loop.